## Importando bibliotecas

In [5]:
import pandas as pd

In [6]:
dados = pd.read_csv(r"C:\Users\Cauan\Desktop\Projetos Finanças (Data Analyst)\Projeto 01 Credit Risk(ETL, EDA)\Dados\credit_risk_dataset.csv")
dados.head(10)

,person_age,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length
0,22,59000,RENT,123.0,PERSONAL,D,35000,16.02,1,0.59,Y,3
1,21,9600,OWN,5.0,EDUCATION,B,1000,11.14,0,0.10,N,2
2,25,9600,MORTGAGE,1.0,MEDICAL,C,5500,12.87,1,0.57,N,3
3,23,65500,RENT,4.0,MEDICAL,C,35000,15.23,1,0.53,N,2
4,24,54400,RENT,8.0,MEDICAL,C,35000,14.27,1,0.55,Y,4
5,21,9900,OWN,2.0,VENTURE,A,2500,7.14,1,0.25,N,2
6,26,77100,RENT,8.0,EDUCATION,B,35000,12.42,1,0.45,N,3
7,24,78956,RENT,5.0,MEDICAL,B,35000,11.11,1,0.44,N,4
8,24,83000,RENT,8.0,PERSONAL,A,35000,8.90,1,0.42,N,2
9,21,10000,OWN,6.0,VENTURE,D,1600,14.74,1,0.16,N,3


In [7]:
dados.columns = 'idade','renda_anual','situacao_moradia','tempo_emprego_anos','motivo_emprestimo','nota_risco','valor_emprestimo','taxa_juros','inadimplencia', 'comprometimento_renda%','historico_inadimplencia','anos_historico_credito'

In [8]:
dados.describe()

,idade,renda_anual,tempo_emprego_anos,valor_emprestimo,taxa_juros,inadimplencia,comprometimento_renda%,anos_historico_credito
count,32581.000000,3.258100e+04,31686.000000,32581.000000,29465.000000,32581.000000,32581.000000,32581.000000
mean,27.734600,6.607485e+04,4.789686,9589.371106,11.011695,0.218164,0.170203,5.804211
std,6.348078,6.198312e+04,4.142630,6322.086646,3.240459,0.413006,0.106782,4.055001
min,20.000000,4.000000e+03,0.000000,500.000000,5.420000,0.000000,0.000000,2.000000
25%,23.000000,3.850000e+04,2.000000,5000.000000,7.900000,0.000000,0.090000,3.000000
50%,26.000000,5.500000e+04,4.000000,8000.000000,10.990000,0.000000,0.150000,4.000000
75%,30.000000,7.920000e+04,7.000000,12200.000000,13.470000,0.000000,0.230000,8.000000
max,144.000000,6.000000e+06,123.000000,35000.000000,23.220000,1.000000,0.830000,30.000000


In [9]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32581 entries, 0 to 32580
Data columns (total 12 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   idade                    32581 non-null  int64  
 1   renda_anual              32581 non-null  int64  
 2   situacao_moradia         32581 non-null  object 
 3   tempo_emprego_anos       31686 non-null  float64
 4   motivo_emprestimo        32581 non-null  object 
 5   nota_risco               32581 non-null  object 
 6   valor_emprestimo         32581 non-null  int64  
 7   taxa_juros               29465 non-null  float64
 8   inadimplencia            32581 non-null  int64  
 9   comprometimento_renda%   32581 non-null  float64
 10  historico_inadimplencia  32581 non-null  object 
 11  anos_historico_credito   32581 non-null  int64  
dtypes: float64(3), int64(5), object(4)
memory usage: 3.0+ MB


##  Trocando o tipo da coluna para poder alterar 1 e 0 para inadimplente e não inadimplente

In [10]:
dados["inadimplencia"] = dados["inadimplencia"].astype(str)

In [11]:
traducao1 = {
            "RENT": "Aluguel",
            "OWN":"Própria",
            "MORTGAGE":"Financiada",
            "OTHER":"Outros"
}
dados["situacao_moradia"] = dados["situacao_moradia"].map(traducao1)

traducao2 = {
        'PERSONAL':"Pessoal",
        'EDUCATION':"Educação", 
        'MEDICAL':"Saúde", 
        'VENTURE':"Empreendimento", 
        'HOMEIMPROVEMENT':"Reforma",
        'DEBTCONSOLIDATION': "Consolidação de Dividas" 
}
dados["motivo_emprestimo"] = dados["motivo_emprestimo"].map(traducao2)



In [12]:
traducao3 = {
    "1":"Inadimplente",
    "0":"Não Inadimplente"    
}

dados["inadimplencia"] = dados["inadimplencia"].map(traducao3)

In [13]:
traducao4 = {
    "Y":"S",
    "N":"N"
}

dados["historico_inadimplencia"] = dados["historico_inadimplencia"].map(traducao4)

In [15]:
dados.head()

,idade,renda_anual,situacao_moradia,tempo_emprego_anos,motivo_emprestimo,nota_risco,valor_emprestimo,taxa_juros,inadimplencia,comprometimento_renda%,historico_inadimplencia,anos_historico_credito
0,22,59000,Aluguel,123.0,Pessoal,D,35000,16.02,Inadimplente,0.59,S,3
1,21,9600,Própria,5.0,Educação,B,1000,11.14,Não Inadimplente,0.10,N,2
2,25,9600,Financiada,1.0,Saúde,C,5500,12.87,Inadimplente,0.57,N,3
3,23,65500,Aluguel,4.0,Saúde,C,35000,15.23,Inadimplente,0.53,N,2
4,24,54400,Aluguel,8.0,Saúde,C,35000,14.27,Inadimplente,0.55,S,4


    Ajustando registros

    Preenchendo os dados nulos da coluna tempo_emprego_anos com a mediana (decidi preencher dessa maneira por causa de alguns possiveis valores discrepantes)

In [16]:
#Arrumando um erro de digitação no primeiro registro da coluna tempo_emprego_anos
dados.iloc[0,3] = 4.0

In [17]:
dados["tempo_emprego_anos"] = dados["tempo_emprego_anos"].fillna(dados["tempo_emprego_anos"].median())

In [18]:
dados.head()

,idade,renda_anual,situacao_moradia,tempo_emprego_anos,motivo_emprestimo,nota_risco,valor_emprestimo,taxa_juros,inadimplencia,comprometimento_renda%,historico_inadimplencia,anos_historico_credito
0,22,59000,Aluguel,4.0,Pessoal,D,35000,16.02,Inadimplente,0.59,S,3
1,21,9600,Própria,5.0,Educação,B,1000,11.14,Não Inadimplente,0.10,N,2
2,25,9600,Financiada,1.0,Saúde,C,5500,12.87,Inadimplente,0.57,N,3
3,23,65500,Aluguel,4.0,Saúde,C,35000,15.23,Inadimplente,0.53,N,2
4,24,54400,Aluguel,8.0,Saúde,C,35000,14.27,Inadimplente,0.55,S,4


In [19]:
#Mudando o tipo de dado da coluna até por que não faz sentido uma coluna de anos ser um float
dados["tempo_emprego_anos"] = dados["tempo_emprego_anos"].astype("int64")

In [20]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32581 entries, 0 to 32580
Data columns (total 12 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   idade                    32581 non-null  int64  
 1   renda_anual              32581 non-null  int64  
 2   situacao_moradia         32581 non-null  object 
 3   tempo_emprego_anos       32581 non-null  int64  
 4   motivo_emprestimo        32581 non-null  object 
 5   nota_risco               32581 non-null  object 
 6   valor_emprestimo         32581 non-null  int64  
 7   taxa_juros               29465 non-null  float64
 8   inadimplencia            32581 non-null  object 
 9   comprometimento_renda%   32581 non-null  float64
 10  historico_inadimplencia  32581 non-null  object 
 11  anos_historico_credito   32581 non-null  int64  
dtypes: float64(2), int64(5), object(5)
memory usage: 3.0+ MB


In [21]:
dados.isnull().sum()

idade                         0
renda_anual                   0
situacao_moradia              0
tempo_emprego_anos            0
motivo_emprestimo             0
nota_risco                    0
valor_emprestimo              0
taxa_juros                 3116
inadimplencia                 0
comprometimento_renda%        0
historico_inadimplencia       0
anos_historico_credito        0
dtype: int64

In [22]:
porcentagem_nulos = round(dados.isnull().mean() * 100,2)

print(f"Porcentagem de dados nulos:",porcentagem_nulos["taxa_juros"])



Porcentagem de dados nulos: 9.56


In [23]:
dados.head()

,idade,renda_anual,situacao_moradia,tempo_emprego_anos,motivo_emprestimo,nota_risco,valor_emprestimo,taxa_juros,inadimplencia,comprometimento_renda%,historico_inadimplencia,anos_historico_credito
0,22,59000,Aluguel,4,Pessoal,D,35000,16.02,Inadimplente,0.59,S,3
1,21,9600,Própria,5,Educação,B,1000,11.14,Não Inadimplente,0.10,N,2
2,25,9600,Financiada,1,Saúde,C,5500,12.87,Inadimplente,0.57,N,3
3,23,65500,Aluguel,4,Saúde,C,35000,15.23,Inadimplente,0.53,N,2
4,24,54400,Aluguel,8,Saúde,C,35000,14.27,Inadimplente,0.55,S,4


> **💡 Decisão de Negócio (Tratamento de Nulos):** Em vez de utilizar a média global e distorcer a regra de precificação da instituição financeira, os 9% de dados faltantes na `taxa_juros` foram preenchidos utilizando a média de agrupamentos baseados no risco do cliente (Cruzamento de Inadimplência Atual x Histórico de Inadimplência). Isso preserva a variância natural onde clientes mais arriscados recebem taxas maiores.

In [24]:
grupo_preenche_nulo = dados.groupby(["inadimplencia","historico_inadimplencia"])["taxa_juros"].transform("mean")

dados["taxa_juros"] = dados["taxa_juros"].fillna(grupo_preenche_nulo)

dados["taxa_juros"] = round(dados["taxa_juros"],2)




In [25]:
dados.isnull().sum()

idade                      0
renda_anual                0
situacao_moradia           0
tempo_emprego_anos         0
motivo_emprestimo          0
nota_risco                 0
valor_emprestimo           0
taxa_juros                 0
inadimplencia              0
comprometimento_renda%     0
historico_inadimplencia    0
anos_historico_credito     0
dtype: int64

In [26]:
print("Porcentagem de nulos na Taxa de Juros após o preenchimento:")
print(dados['taxa_juros'].isnull().mean() * 100)

Porcentagem de nulos na Taxa de Juros após o preenchimento:
0.0


In [27]:
dados["renda_mensal"] = round(dados["renda_anual"] / 12,0)

limites_idade = [0, 25, 35, 45, 60, 120]
nomes_faixas = ['Até 25 anos', '26-35 anos', '36-45 anos', '46-60 anos', 'Mais de 60 anos']
dados['faixa_etaria'] = pd.cut(dados['idade'], bins=limites_idade, labels=nomes_faixas)

In [28]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32581 entries, 0 to 32580
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype   
---  ------                   --------------  -----   
 0   idade                    32581 non-null  int64   
 1   renda_anual              32581 non-null  int64   
 2   situacao_moradia         32581 non-null  object  
 3   tempo_emprego_anos       32581 non-null  int64   
 4   motivo_emprestimo        32581 non-null  object  
 5   nota_risco               32581 non-null  object  
 6   valor_emprestimo         32581 non-null  int64   
 7   taxa_juros               32581 non-null  float64 
 8   inadimplencia            32581 non-null  object  
 9   comprometimento_renda%   32581 non-null  float64 
 10  historico_inadimplencia  32581 non-null  object  
 11  anos_historico_credito   32581 non-null  int64   
 12  renda_mensal             32581 non-null  float64 
 13  faixa_etaria             32576 non-null  category
dtypes: cat

In [29]:
dados["renda_mensal"] = dados["renda_mensal"].astype("Int64")

Identificando quais são os outliers que estão gerando 5 nulos na coluna faixa etaria

In [30]:
linhas_com_nulo = dados[dados["faixa_etaria"].isnull()]

print(linhas_com_nulo)

       idade  renda_anual situacao_moradia  tempo_emprego_anos  \
81       144       250000          Aluguel                   4   
183      144       200000       Financiada                   4   
575      123        80004          Aluguel                   2   
747      123        78000          Aluguel                   7   
32297    144      6000000       Financiada                  12   

      motivo_emprestimo nota_risco  valor_emprestimo  taxa_juros  \
81       Empreendimento          C              4800       13.57   
183            Educação          B              6000       11.86   
575            Educação          B             20400       10.25   
747      Empreendimento          B             20000        9.84   
32297           Pessoal          C              5000       12.73   

          inadimplencia  comprometimento_renda% historico_inadimplencia  \
81     Não Inadimplente                    0.02                       N   
183    Não Inadimplente                    0

Durante a checagem das novas colunas, identifiquei 5 registros com idades irreais (entre 123 e 144 anos), caracterizando erros de digitação ou falhas no sistema legado. Como representavam uma parcela ínfima da base, optei pela exclusão (drop) dessas linhas para não enviesar as faixas etárias.

In [31]:
#Conferindo o resultado
dados = dados[dados['idade'] <= 100]
print(dados[dados['idade'] > 100])

Empty DataFrame
Columns: [idade, renda_anual, situacao_moradia, tempo_emprego_anos, motivo_emprestimo, nota_risco, valor_emprestimo, taxa_juros, inadimplencia, comprometimento_renda%, historico_inadimplencia, anos_historico_credito, renda_mensal, faixa_etaria]
Index: []


In [32]:
dados.head()

,idade,renda_anual,situacao_moradia,tempo_emprego_anos,motivo_emprestimo,nota_risco,valor_emprestimo,taxa_juros,inadimplencia,comprometimento_renda%,historico_inadimplencia,anos_historico_credito,renda_mensal,faixa_etaria
0,22,59000,Aluguel,4,Pessoal,D,35000,16.02,Inadimplente,0.59,S,3,4917,Até 25 anos
1,21,9600,Própria,5,Educação,B,1000,11.14,Não Inadimplente,0.10,N,2,800,Até 25 anos
2,25,9600,Financiada,1,Saúde,C,5500,12.87,Inadimplente,0.57,N,3,800,Até 25 anos
3,23,65500,Aluguel,4,Saúde,C,35000,15.23,Inadimplente,0.53,N,2,5458,Até 25 anos
4,24,54400,Aluguel,8,Saúde,C,35000,14.27,Inadimplente,0.55,S,4,4533,Até 25 anos


In [33]:
dados.to_csv("dados_tratados.csv",index=False)